### ISSUE TITLE
Inference should flow information between arguments in a generic function call

### ISSUE THREAD

**Issue by leafpetersen** [dart-lang/language#731]
EDIT by @jmesserly

Inference should flow information up/down between arguments in a generic function call. This will help inference common methods like `Iterable.fold` and `Future.forEach`.

----

(original description)
We currently do downwards and upwards inference independently.  This gets us a fair bit, but we should eventually integrate the two, probably following the colored local type inference approach.

**Comment by mnordine** [comment]
@leafpetersen Is this more of an issue now with nnbd? 

Somebody posted the following example in Gitter, complaining of the lack of inference. I tried with nnbd and it results in a static error:

```dart
Iterable<R> map<T, R>(
  Iterable<T> iterable,
  R Function(T) transform,
) sync* {
  for (var item in iterable) {
    yield transform(item);
  }
}

void main() {
  var list = [1, 2, 3];
  var mappedList = map(list, (x) => x + 1); // `x` is inferred as `Object?`
  print(mappedList);
}
```

`An expression whose value can be 'null' must be null-checked before it can be dereferenced - line 12`

**Comment by leafpetersen** [comment]
> @leafpetersen Is this more of an issue now with nnbd?

Yes, or no, depending how you look at it.  nnbd will make this more statically visible, which has both upsides and downsides.  

Your example used to work because `x` was typed as dynamic.  This is nice, because the code in fact runs correctly.  But it's also not nice because code that doesn't run correctly will also pass the static analysis.

```dart
void main() {
  var list = [1, 2, 3];
  var mappedList = map(list, (x) => x.foo); // `x` is inferred as `Object?`
  print(mappedList);
}
```

This variant of your `main` function will pass static analysis in pre-nnbd Dart, but blow up at runtime.

But I do expect that nnbd will raise the profile of this issue, yes.

**Comment by kasperpeulen** [comment]
I actually love that nnbd raises an error.
The pre nnbd variant way doesnt look sound at all.

But you will get more people complaining 😅. If something blows up at runtime, the programmer get blamed by their manager. When something doesnt compile that theoretically should, the language designers gets blamed.

**Comment by mnordine** [comment]
Sure, and I'd love inferring the anonymous function parameter even more, :)

**Comment by escamoteur** [comment]
I ran today into this too:
![image](https://user-images.githubusercontent.com/16635729/95088608-71d58300-0723-11eb-992d-47bba692ca1c.png)

**Comment by lulupointu** [comment]
Hey, just wanted to share my experience with regard to this issue.

**Why I use generics**
I built a reusable widget tree in flutter + reusable bloc, which can be used to show information based on the generic being passed. The generic is an object which can be of type `Something` or `SomethingDetailed`. This is quite powerful in my case since both the bloc and the widget tree would be copy paste of the other, apart from some `Chips` which can be tapped to reveal more information.

**My experience relative to this issue**
All things considered I am really happy with the result. However the experience of reusing this object was sometimes painful because (and this is my biggest complain) the compiler errors where so off. Many time have I had the error **`SomethingDetailed` is not of type `Detailed`** (which is it). The true error was that I failed to specify the generic somewhere. As you might imagine this is really hard to debug. 

`implicit-dynamic: false` does help but is far from perfect. I had the issue described above without the lint being triggered.

Anyway I did encounter other situations where having this feature would have help but it was more about convenience. This experience here was a real pain so I wanted to share it :wink:

**Comment by leafpetersen** [comment]
> I built a reusable widget tree in flutter + reusable bloc, which can be used to show information based on the generic being passed.

@lulupointu Can you share your code or extract a snippet that demonstrates the exact pattern you're running into?  @stereotype441 is looking into this and is trying to understand what patterns show up in real code.

**Comment by jodinathan** [comment]
would this snippet work?

```dart
void foo<T extends Object>(T bar, Function(T) fn) {
  
}

void main() {
  foo('fooo', (zz) {
    print(zz.length); // zz is Object? instead of String
  });
}

```

would the `zz` variable be infered as `String` with this feature?

**Comment by leafpetersen** [comment]
> would the `zz` variable be infered as `String` with this feature?

Yes.

**Comment by jodinathan** [comment]
this is so nice @leafpetersen!  

I wonder if is this small enough to come before greater ones like macros, patterns etc?

**Comment by leafpetersen** [comment]
> I wonder if is this small enough to come before greater ones like macros, patterns etc?

No promises, but current plan is for this and some other small inference/promotion improvements to roll out this summer.

**Comment by mnordine** [comment]
@stereotype441 I have the latest sdk installed from `main`, added the following to `analysis_options.yaml`:

```yaml
enable-experiment:
    - inference-update-1
```

And I still get this:
<img width="364" alt="object" src="https://user-images.githubusercontent.com/1924573/159709892-86c0ca74-caec-4bdd-84d6-774572ae8949.png">

Anything else I need to do? Or does this not work yet?

**Comment by stereotype441** [comment]
@mnordine I love your enthusiasm!  Unfortunately I'm still in the pretty early stages of developing this feature, so it doesn't work yet.  I'm going to work on the analyzer support first, and then add support in the compiler front end.

I'll post an update when I've done enough of the implementation that it's worth trying.

**Comment by stereotype441** [comment]
Hi @mnordine (and anyone else eagerly following along).  As of https://github.com/dart-lang/sdk/commit/83bea5cfbb40029bcbb3a576ba2a585e441a55d2, I have added preliminary support to the analyzer for this feature.  Feel free to check it out and let me know what you think!  A few caveats:
- The code only handles simple cases where the argument types of closures get inferred from the types of non-closure arguments.  I still intend to do some follow-up work to handle more complex cases where the argument types of one closure get inferred from the type of a different closure.
- You'll have to turn on the experimental flag `inference-update-1` to see any effect.
- The feature will only work in files whose language version is `2.17`.
- The feature is only implemented in the analyzer so far.  So you'll be able to play with the feature in an editor like IntelliJ or VS Code (or using `dart analyze`), but you won't be able to run a program that takes advantage of it.
- You'll either need to be on the dev channel (version `2.17.0-243.0.dev` or later) or build the latest version of the Dart SDK yourself.
- We aren't making any claims about compatibility yet.  It's possible that we'll still discover an issue in the design and that we'll decide to change how the feature works, in a way that might break a program that relies on it.

**Comment by pedromassango** [comment]
@felangel can this prevent the need to specify the state when declaring a BlocBuilder widget?

From (current): `BlocBuilder<MyCubit, MyState>(...)`
To (new/wanted): `BlocBuilder<MyCubit>(...)`

**Comment by felangel** [comment]
> @felangel can this prevent the need to specify the state when declaring a BlocBuilder widget?
> 
> From (current): `BlocBuilder<MyCubit, MyState>(...)` To (new/wanted): `BlocBuilder<MyCubit>(...)`

Looks like it might! I’ll check shortly 👍

**Comment by eernstg** [comment]
I think it would be useful to clarify the situation around [this comment](https://github.com/dart-lang/language/issues/731#issuecomment-1164795701). Here is a simpler example showing the same phenomenon as the example given there:

```dart
X f<X>(X x, X Function(X) g) => g(x);

void main() {
  var i = 1;
  num n = f(i, (j) => j.bitLength); // 'The getter `bitLength` isn't defined for the type `num`'.
}
```

In this case (and in the original example) there is no direct connection to the support for horizontal inference (which is the new feature that was introduced to resolve this issue). The reason why the inferred actual type argument to the invocation of `f` is `num` rather than `int` is that the context type is `num`. That is, this inference result is caused by declaring `n` as `num n`, rather than declaring it, e.g., as `var n` or `int n`.

This means that there is no reasonable way we could change horizontal inference such that these examples would infer `int` rather than `num`. So there is no reason to have that discussion here.

It is a separate discussion that Dart type inference gives priority to the context type, rather than choosing a type which is obtained by plain bottom-up inference. Check out the [issues about inference in general](https://github.com/dart-lang/language/issues?q=is%3Aissue+is%3Aopen+label%3Ainference+), which currently shows at least four issues about how to handle context types. The standard example showing that the context type cannot be ignored in Dart could be this one:

```dart
// Assume that Dart would give priority to bottom-up inference,
// rather than giving priority to the context type.

void main() {
  List<num> nums = []; // Inferred as `<Never>[]`.
  nums.add(1); // Throws.
}
```

**Comment by Miracle-Blue** [comment]
```dart:
